joint_trajectory_controller/src/trajectory.cpp

```diff
@@ -88,17 +88,56 @@ bool Trajectory::sample(
     return false;
   }
 
+  auto deduce_from_derivatives = [&](
+                                   trajectory_msgs::msg::JointTrajectoryPoint & first_state,
+                                   trajectory_msgs::msg::JointTrajectoryPoint & second_state,
+                                   const size_t dim, const double delta_t) {
+    if (second_state.positions.empty())
+    {
+      second_state.positions.resize(dim);
+      if (first_state.velocities.empty())
+      {
+        first_state.velocities.resize(dim, 0.0);
+      }
+      if (second_state.velocities.empty())
+      {
+        second_state.velocities.resize(dim);
+        if (first_state.accelerations.empty())
+        {
+          first_state.accelerations.resize(dim, 0.0);
+        }
+        for (size_t i = 0; i < dim; ++i)
+        {
+          second_state.velocities[i] =
+            first_state.velocities[i] +
+            (first_state.accelerations[i] + second_state.accelerations[i]) * 0.5 * delta_t;
+        }
+      }
+      for (size_t i = 0; i < dim; ++i)
+      {
+        // second state velocity should be reached on the end of the segment, so use middle
+        second_state.positions[i] =
+          first_state.positions[i] +
+          (first_state.velocities[i] + second_state.velocities[i]) * 0.5 * delta_t;
+      }
+    }
+  };
+
   // current time hasn't reached traj time of the first point in the msg yet
-  const auto & first_point_in_msg = trajectory_msg_->points[0];
-  const rclcpp::Duration offset = first_point_in_msg.time_from_start;
-  const rclcpp::Time first_point_timestamp = trajectory_start_time_ + offset;
+  auto & first_point_in_msg = trajectory_msg_->points[0];
+  const rclcpp::Time first_point_timestamp =
+    trajectory_start_time_ + first_point_in_msg.time_from_start;
+
   if (sample_time < first_point_timestamp)
   {
-    const rclcpp::Time t0 = time_before_traj_msg_;
+    deduce_from_derivatives(
```

Comment:
This is not needed since the method will be only executed if `positions` and/or `velocities` are empty and their derivative exists. (And we can save a parameter in arguments)

Reply:
:+1: